libs/contrib/nanovg_helper.xtm

```diff
@@ -0,0 +1,742 @@
+;;; nanovg_helper.xtm -- nanovg helper library
+
+;; Author: Yuya Chiba (cy.blue.9@gmail.com)
+;; Keywords: extempore
+;; Required dylibs: libnanovg
+
+;;; Commentary:
+
+;; This library is warrper library of nanovg(libs/external/nanovg.xtm)
+
+;;; Code:
+
+(sys:load "libs/aot-cache/nanovg_helper.xtm" 'quiet)
+(sys:load-preload-check 'nanovg_helper)
+(define *xtmlib-nanovg_helper-loaded* #f)
+
+(impc:aot:suppress-aot-do
+ (sys:load "libs/external/nanovg.xtm"))
+(impc:aot:insert-forms
+ (sys:load "libs/external/nanovg.xtm" 'quiet))
+
+(impc:aot:insert-header "xtmnanovg_helper")
+
+;; set up the current dylib name and path (for AOT compilation)
+(bind-dylib libnanovg
+  (cond ((string=? (sys:platform) "OSX")
+         "libnanovg.dylib")
+        ((string=? (sys:platform) "Linux")
+         "libnanovg.so")
+        ((string=? (sys:platform) "Windows")
+         "nanovg.dll")))
+
+;;;;;;;;;;;;;;;;;;;;
+;; header library ;;
+;;;;;;;;;;;;;;;;;;;;
+
+(bind-func xtm_stroke
+  (lambda (path color:NVGcolor* width:i64)
+    (nvgStrokeWidth path (i64tof width))
+    (nvgStrokeColor path color)
+    (nvgStroke path)))
+
+(bind-func xtm_stroke
+  (lambda (path paint:NVGpaint* width:i64)
+    (nvgStrokeWidth path (i64tof width))
+    (nvgStrokePaint path paint)
+    (nvgStroke path)))
+
+(bind-func xtm_stroke
+  (lambda (path color:NVGcolor* width:float)
+    (nvgStrokeWidth path width)
+    (nvgStrokeColor path color)
+    (nvgStroke path)))
+
+(bind-func xtm_stroke
+  (lambda (path paint:NVGpaint* width:float)
+    (nvgStrokeWidth path width)
+    (nvgStrokePaint path paint)
+    (nvgStroke path)))
+
+(bind-func xtm_fill
+  (lambda (path color:NVGcolor*)
+    (nvgFillColor path color)
+    (nvgFill path)))
+
+(bind-func xtm_fill
+  (lambda (path paint:NVGpaint*)
+    (nvgFillPaint path paint)
+    (nvgFill path)))
+
+;; convenience functions for adding shapes to path
+
+(bind-func xtm_add_line
```

Comment:
these wrapper functions are all fine, but for your own convenience you probably don't need to add all the explicit type annotations to the function arguments---the type inferencer should figure out most of them.

It's fine to leave them as-is, though.

Reply:
Thanks for advice.
You are right.

But, this helper library provides two pattern interface (`i64` type or `float` type) for each function to be easy writing code.
For example,
```
(bind-func xtm_add_line
  (lambda (path x1:i64 y1:i64 x2:i64 y2:i64)
    (nvgMoveTo path (i64tof x1) (i64tof y1))
    (nvgLineTo path (i64tof x2) (i64tof y2))))

(add_line vg 0 0 100 100)


(bind-func xtm_add_line
  (lambda (path x1:float y1:float x2:float y2:float)
    (nvgMoveTo path x1 y1)
    (nvgLineTo path x2 y2)))

(add_line 0. 0. 100. 100.)
```
Above two examples draw a same line.

If the explicit type annotations to the function arguments is nothing, the programmer who use this libraliy for the first time may confuse the difference between the above two functions.
So, I explicited type annotations to the function arguments. (Probably because my xtlang experience is not enough...)

Tentatively, I leave it.